File: src/m/plot/processmesh.py
#import matplotlib.delaunay as delaunay
import numpy as np


def processmesh(md, data, options):
    """PROCESSMESH - process mesh to be plotted

    Usage:
        x, y, z, elements, is2d = processmech(md, data, options)

    See also: PLOTMODEL, PROCESSDATA

    TODO:
    - Test application of matlplotlib.delaunay
    - Test that output of delaunay matches output of delaunay in MATLAB (see
    src/m/plot/processmesh.m)
    """

    #some checks
    if md.mesh.numberofvertices == 0:
        raise ValueError('processmesh error: mesh is empty')
    if md.mesh.numberofvertices == md.mesh.numberofelements:
        raise ValueError('processmesh error: the number of elements is the same as the number of nodes')

    #special case for mesh 2dvertical
    if md.mesh.domaintype() == '2Dvertical':
        x, y, z, elements, is2d, isplanet = processmesh(md.mesh, options)
        return x, y, z, elements, is2d, isplanet

    # # special case for mesh 3dsurface
    # if md.mesh.domaintype() == '3Dsurface':
    #     x, y, z, elements, is2d, isplanet = processmesh(md.mesh, options)
    #     if options.getfieldvalue('coord', 'xy') == 'latlon':
    #         x = md.mesh.long
    #         y = md.mesh.lat
    #         elements = delaunay(x, y)
    #         z = md.mesh.lat
    #         z[:] = 0
    #     return x, y, z, elements, is2d, isplanet

    if hasattr(md.mesh, 'elements2d'):
        elements2d = md.mesh.elements2d
        numofvertices2d = md.mesh.numberofvertices2d
        numofelements2d = md.mesh.numberofelements2d
    else:
        numofvertices2d = np.nan
        numofelements2d = np.nan

    if options.exist('amr'):
        step = options.getfieldvalue('amr')
        nonan = np.nonzero(~np.isnan(md.results.TransientSolution[step].MeshX))
        x = md.results.TransientSolution[step].MeshX[nonan]
        y = md.results.TransientSolution[step].MeshY[nonan]
        nonan = np.nonzero(~np.isnan(md.results.TransientSolution[step].MeshElements))
        elements = md.results.TransientSolution[step].MeshElements[nonan] - 1
        eldim = np.shape(md.results.TransientSolution[step].MeshElements)[1]
        elements = np.reshape(elements, ((int(len(elements) / eldim), eldim)))

    else:
        elements = md.mesh.elements - 1
        if options.getfieldvalue('coord', 'xy') != 'latlon':
            x = md.mesh.x
            if hasattr(md.mesh, 'x2d'):
                x2d = md.mesh.x2d
            y = md.mesh.y
            if hasattr(md.mesh, 'y2d'):
                y2d = md.mesh.y2d

    if hasattr(md.mesh, 'z'):
        z = md.mesh.z
    else:
        z = np.zeros(np.shape(md.mesh.x))
    z = options.getfieldvalue('z', z)
    if isinstance(z, str):
        z = getattr(md, z)

    force2D = numofelements2d in np.shape(data) or numofvertices2d in np.shape(data)
    #is it a 2D plot?
    if md.mesh.dimension() == 2 or options.getfieldvalue('layer', 0) >= 1 or force2D:
        is2d = 1
    else:
        is2d = 0

    #layer projection?
    if options.getfieldvalue('layer', 0) >= 1 or force2D:
        if options.getfieldvalue('coord', 'xy') == 'latlon':
            raise Exception('processmesh error message: cannot work with 3D meshes for now')
        #we modify the mesh temporarily to a 2d mesh from which the 3d mesh was extruded
        x = x2d
        y = y2d
        z = np.zeros(np.shape(x2d))
        elements = md.mesh.elements2d - 1

    #units
    if options.exist('unit'):
        unit = options.getfieldvalue('unit')
        x = x * unit
        y = y * unit
        z = z * unit

    #is model a member of planet class?
    #
    # TODO: Change this when planet class defined (see src/m/plot/processmesh.m)
    #
    if md.__class__.__name__ != 'model':
        isplanet = 1
    else:
        isplanet = 0

    return x, y, z, elements, is2d, isplanet

File: src/m/plot/processmesh.m
function [x y z elements is2d isplanet]=processmesh(md,data,options)
%PROCESSMESH - process mesh to be plotted
%
%   Usage:
%      [x y z elements is2d]=processmesh(md,data,options)
%
%   See also: PLOTMODEL, PROCESSDATA

%some checks
if md.mesh.numberofvertices==0,
	error('plot error message: mesh is empty')
end
if md.mesh.numberofvertices==md.mesh.numberofelements
	error(['plot error message: the number of elements is the same as the number of nodes...']);
end

%special case for mesh 2dvertical
if strcmp(domaintype(md.mesh),'2Dvertical'),
	[x y z elements is2d isplanet] = processmesh(md.mesh,options);
	return;
end

%special case for mesh 3dsurface
if strcmp(domaintype(md.mesh),'3Dsurface'),
	[x y z elements is2d isplanet] = processmesh(md.mesh,options);
	if strcmpi(getfieldvalue(options,'coord','xy'),'latlon') | strcmpi(getfieldvalue(options,'coord','xy'),'latlong'),
		x0=md.mesh.long;
		y0=md.mesh.lat;
		%add row at lat=90 and lat=-90
		add=[(-180:.1:-1e-5)';(1e-5:.1:180)'];
		nadd=length(add);
		xextra=[add;add];
		yextra=[90*ones(nadd,1); -90*ones(nadd,1)];
		x=[x0;xextra];
		y=[y0;yextra];

		if strcmpi(getfieldvalue(options,'coordcent','atlantic'),'pacific'),
			pos=find(x>0);  x(pos)=-360+x(pos);
		end
		elements=delaunay(x,y);
		z=x; z(:)=0;
	end
	return;
end

if isprop(md.mesh,'elements2d'), elements2d=md.mesh.elements2d; end

if exist(options,'amr'),
	step = getfieldvalue(options,'amr');
	x = md.results.TransientSolution(step).MeshX;
	y = md.results.TransientSolution(step).MeshY;
	elements = md.results.TransientSolution(step).MeshElements;
else
	elements=md.mesh.elements;
	if ~strcmpi(getfieldvalue(options,'coord','xy'),'latlon') &  ~strcmpi(getfieldvalue(options,'coord','xy'),'latlong') ,
		x=md.mesh.x;
		if isprop(md.mesh,'x2d'), x2d=md.mesh.x2d; end
		y=md.mesh.y;
		if isprop(md.mesh,'y2d'), y2d=md.mesh.y2d; end
	else
		x=md.mesh.long;
		y=md.mesh.lat;
		if strcmpi(getfieldvalue(options,'coordcent','atlantic'),'pacific'),
			pos=find(x>0);  x(pos)-360+x(pos);
		end
	end
end

if isprop(md.mesh,'z'),
	z=md.mesh.z;
else
	z=zeros(size(x));
end
z=getfieldvalue(options,'z',z);
if ischar(z),
	z=md.(z);
end

%is it a 2d plot?
if md.mesh.dimension()==2 || getfieldvalue(options,'layer',0)>=1 || getfieldvalue(options,'depthaverage',0)
	is2d=1;
else
	is2d=0;
end

%layer projection? 
if getfieldvalue(options,'layer',0)>=1 || getfieldvalue(options,'depthaverage',0)
	if strcmpi(getfieldvalue(options,'coord','xy'),'latlon'),
		error('processmesh error message: cannot work with 3D meshes for now');
	end
	%we modify the mesh temporarily to a 2d mesh from which the 3d mesh was extruded
	x=x2d;
	y=y2d;
	z=zeros(size(x2d));
	elements=elements2d;
end

%units
if exist(options,'unit'),
	unit=getfieldvalue(options,'unit');
	x=x*unit;
	y=y*unit;
	z=z*unit;
end

%Quiver plot for elements?
if size(data,2)>1 && size(data,1)==size(elements,1)
	x = mean(x(elements),2);
	y = mean(y(elements),2);
end

if isa(md,'planet'),
	isplanet=1;
else
	isplanet=0;
end
